File: src/com/dalolorn/sr2modmanager/MainController.java
package com.dalolorn.sr2modmanager;

import javafx.application.Platform;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.VBox;
import javafx.stage.DirectoryChooser;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.stage.StageStyle;
import org.eclipse.jgit.api.CreateBranchCommand;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.ListBranchCommand;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.errors.RepositoryNotFoundException;
import org.eclipse.jgit.lib.*;
import org.eclipse.jgit.treewalk.TreeWalk;
import org.eclipse.jgit.util.FileUtils;

import java.io.*;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

public class MainController {
	@FXML private TextField urlField;
	@FXML private Label urlLabel;
	@FXML private TextArea modInfo, branchInfo;
	@FXML private ListView<String> branchList;

	Git repo;

	Ref currentBranch;
	HashMap<String, Ref> branches = new HashMap<>();
	String gamePath = "..";

	public void initialize() {
		branchList.getSelectionModel().selectedItemProperty().addListener(new ChangeListener<String>() {
			@Override
			public void changed(ObservableValue<? extends String> observable, String oldValue, String newValue) {
				setActiveBranch(newValue);
			}
		});
	}

	private void setActiveBranch(String branchName) {
		currentBranch = branches.get(branchName);

		ObjectLoader descriptionLoader;
		try {
			Repository repository = repo.getRepository();
			ObjectId treeId = repository.resolve(currentBranch.getName() + "^{tree}");
			TreeWalk walker = Utils.generateBranchDescWalker(repository, treeId);
			ObjectId descriptionId = walker.getObjectId(0);
			descriptionLoader = repo.getRepository().open(descriptionId);
		} catch (Exception e) {
			branchInfo.setText("No description could be found for this branch.");
			e.printStackTrace();
			return;
		}

		try (
				InputStream descriptionStream = descriptionLoader.openStream();
				BufferedReader descriptionReader = new BufferedReader(new InputStreamReader(descriptionStream));
		) {
			branchInfo.setText(descriptionReader.lines().collect(Collectors.joining("\n")));
		} catch (NullPointerException | IOException e) {
			branchInfo.setText("No description could be found for this branch.");
		}
	}

	@FXML private void connectToRepository(ActionEvent event) {
		try {
			// Get and parse URL, prepare repository root.
			String url = urlField.getText();
			if(!url.startsWith("http"))
				url = "https://" + url;
			if(!url.endsWith(".git"))
				url += ".git";
			String[] path = url.split("/");
			final String repoURL = url;
			String repoName = path[path.length - 2] + "_" + path[path.length - 1];
			repoName = repoName.substring(0, repoName.length() - 4);
			File tempRoot = new File("repositories" + File.separator + repoName);
			if (!tempRoot.exists() || !tempRoot.isDirectory())
				tempRoot = Files.createDirectories(tempRoot.toPath()).toFile();
			final File localRoot = tempRoot;

			// Start preparing the download task.
			final double wndwWidth = 300.0d;
			Label updateLabel = new Label("Loading repository...");
			updateLabel.setPrefWidth(wndwWidth);
			ProgressBar progress = new ProgressBar();
			progress.setPrefWidth(wndwWidth);

			VBox updatePane = new VBox();
			updatePane.setPadding(new Insets(10));
			updatePane.setSpacing(5.0d);
			updatePane.getChildren().addAll(updateLabel, progress);

			Stage taskUpdateStage = new Stage(StageStyle.UTILITY);
			taskUpdateStage.initModality(Modality.APPLICATION_MODAL);
			taskUpdateStage.setScene(new Scene(updatePane));
			taskUpdateStage.show();

			Task updateTask = new Task<Void>() {
				@Override
				protected Void call() throws Exception {
					try {
						updateMessage("Loading repository...");
						try {
							repo = Git.open(localRoot);
						}
						catch (IOException e) {
							repo = Git.cloneRepository()
									.setURI(repoURL)
									.setDirectory(localRoot)
									.call();
						}

						updateMessage("Fetching branch list...");
						getBranches();

						updateMessage("Getting mod description...");
						getDescription();
					} catch (Exception e) {
						e.printStackTrace();
						Platform.runLater(() -> {
							Alert msg = new Alert(Alert.AlertType.ERROR, "Encountered an exception: " + e.toString());
							msg.show();
						});
						throw e;
					}
					return null;
				}
			};

			updateTask.setOnSucceeded(t -> {
				taskUpdateStage.close();
				urlLabel.setText("Connected to" + repoURL);
			});
			updateTask.setOnFailed(t -> taskUpdateStage.close());
			progress.progressProperty().bind(updateTask.progressProperty());
			updateLabel.textProperty().bind(updateTask.messageProperty());

			taskUpdateStage.show();
			new Thread(updateTask).start();
		}
		catch(Exception e) {
			Alert msg = new Alert(Alert.AlertType.ERROR, "Encountered an exception: " + e.toString());
			e.printStackTrace();
			msg.show();
		}
	}

	@FXML private void installMod(ActionEvent event) {
		try {
			// Check if we've got this branch already.
			boolean createBranch = repo.getRepository().resolve(currentBranch.getName().replaceFirst("refs/remotes/origin", "refs/heads")) == null;

			repo.checkout()
					.setName(currentBranch.getName().replaceFirst("refs/remotes/origin/", ""))
					.setCreateBranch(createBranch)
					.setUpstreamMode(CreateBranchCommand.SetupUpstreamMode.TRACK)
					.setStartPoint(currentBranch.getName())
					.call();

			repo.pull().call();

			File root = repo.getRepository().getWorkTree();

			// Find the folder containing the modinfo file.
			boolean inRoot = false;
			String mod = null;
			Finder finder = new Finder("modinfo.txt");
			Files.walkFileTree(root.getAbsoluteFile().toPath(), finder);
			if(finder.getResult() != null) {
				mod = finder.getResult().getParent().getFileName().toString();
				if(mod.equalsIgnoreCase(root.toPath().getFileName().toString())) {
					inRoot = true; // The modinfo is in the repository root, so we can't discard metadata.
					Alert msg = new Alert(Alert.AlertType.WARNING, "WARNING: Unable to discard repository metadata!\n\nTo improve loading times, it is recommended that you delete the installed mod's .git folder once installation is completed.");
					msg.show();
				}
			}
			else {
				Alert msg = new Alert(Alert.AlertType.ERROR, "Cannot find modinfo.txt!\n\nThis repository does not appear to contain a valid Star Ruler 2 mod.\nPlease make sure that you have connected to the right repository, and contact the mod developer if the issue persists.");
				msg.show();
				return;
			}

			File source;
			if(inRoot)
				source = root;
			else
				source = new File(root + File.separator + mod);
			File destination = new File(gamePath + File.separator + "mods" + File.separator + mod);
			if(destination.exists()) { // Wipe out any previous installation to avoid instances of failed
				try {
					FileUtils.delete(destination, FileUtils.RECURSIVE);
				} catch (IOException e) {
					Alert msg = new Alert(Alert.AlertType.ERROR, "Could not delete previous mod installation!\n\nA possible workaround might be to delete the mod folder yourself, then try again.");
					msg.show();
					return;
				}
			}
			if(destination.exists()) { // We failed to delete the folder, but didn't throw an exception.
				Alert msg = new Alert(Alert.AlertType.ERROR, "Could not delete previous mod installation!\n\nA possible workaround might be to delete the mod folder yourself, then try again.");
				msg.show();
				return;
			}

			destination.mkdirs();
			Files.walkFileTree(source.getAbsoluteFile().toPath(), new CopyFileVisitor(destination.getAbsoluteFile().toPath()));
			Alert msg = new Alert(Alert.AlertType.INFORMATION, "Mod successfully installed!");
			msg.show();
		} catch (Exception e) {
			Alert msg = new Alert(Alert.AlertType.ERROR, "Encountered an exception: " + e.toString());
			e.printStackTrace();
			msg.show();
		}
	}

	private void getBranches() throws GitAPIException {
		repo.fetch().call();
		List<Ref> tags = repo.branchList().setListMode(ListBranchCommand.ListMode.REMOTE).call();
		tags.addAll(repo.tagList().call());

		List<String> tagNames = new ArrayList<>();
		branches.clear();
		for (Ref tag : tags) {
			tagNames.add(tag.getName().replaceFirst("refs/remotes/origin/|refs/tags/", ""));
			branches.put(tag.getName().replaceFirst("refs/remotes/origin/|refs/tags/", ""), tag);
		}

		Platform.runLater(() -> {
			branchList.getItems().clear();
			branchList.getItems().addAll(tagNames);
		});
	}

	private void getDescription() throws GitAPIException, IOException {
		Repository repository = repo.getRepository();
		ObjectId treeId = repository.resolve("master^{tree}");
		TreeWalk walker = Utils.generateReadmeWalker(repository, treeId);
		ObjectId readmeId = walker.getObjectId(0);
		ObjectLoader readmeLoader = repo.getRepository().open(readmeId);

		try (
				InputStream readmeStream = readmeLoader.openStream();
				BufferedReader readmeReader = new BufferedReader(new InputStreamReader(readmeStream));
		)
		{
			modInfo.setText(readmeReader.lines().collect(Collectors.joining("\n")));
		}
		catch(NullPointerException|IOException e) {
			modInfo.setText("No description could be found for this mod.");
		}

		Platform.runLater(() -> branchInfo.setText("No branches selected"));
	}

	@FXML private void openRepository(ActionEvent actionEvent) {
		// Get repository location.
		DirectoryChooser chooser = new DirectoryChooser();
		chooser.setInitialDirectory(new File("repositories/"));
		chooser.getInitialDirectory().mkdirs();
		chooser.setTitle("Choose a Repository");
		final File dir = chooser.showDialog(urlField.getScene().getWindow()); // Anything on the window would suffice, urlField was just arbitrarily selected.
		if(dir == null)
			return;

		// Start preparing the download task.
		final double wndwWidth = 300.0d;
		Label updateLabel = new Label("Loading repository...");
		updateLabel.setPrefWidth(wndwWidth);
		ProgressBar progress = new ProgressBar();
		progress.setPrefWidth(wndwWidth);

		VBox updatePane = new VBox();
		updatePane.setPadding(new Insets(10));
		updatePane.setSpacing(5.0d);
		updatePane.getChildren().addAll(updateLabel, progress);

		Stage taskUpdateStage = new Stage(StageStyle.UTILITY);
		taskUpdateStage.initModality(Modality.APPLICATION_MODAL);
		taskUpdateStage.setScene(new Scene(updatePane));
		taskUpdateStage.show();

		Task updateTask = new Task<Void>() {
			@Override
			protected Void call() throws Exception {
				try {
					updateMessage("Loading repository...");
					repo = Git.open(dir);

					updateMessage("Fetching branch list...");
					getBranches();

					updateMessage("Getting mod description...");
					getDescription();
				}
				catch (RepositoryNotFoundException e) {
					e.printStackTrace();
					Platform.runLater(() -> {
						Alert msg = new Alert(Alert.AlertType.ERROR, "Not a valid repository!");
						msg.show();
					});
					throw e;
				}
				catch (Exception e) {
					e.printStackTrace();
					Platform.runLater(() -> {
						Alert msg = new Alert(Alert.AlertType.ERROR, "Encountered an exception: " + e.toString());
						msg.show();
					});
					throw e;
				}
				return null;
			}
		};

		updateTask.setOnSucceeded(t -> {
			taskUpdateStage.close();

			String url = repo.getRepository().getConfig().getString("remote", "origin", "url");
			if(url == null)
				url = "a local repository at " + dir.getAbsolutePath();

			urlLabel.setText("Connected to " + url);
		});
		updateTask.setOnFailed(t -> taskUpdateStage.close());
		progress.progressProperty().bind(updateTask.progressProperty());
		updateLabel.textProperty().bind(updateTask.messageProperty());

		taskUpdateStage.show();
		new Thread(updateTask).start();
	}

	@FXML private void close(ActionEvent actionEvent) {
		Alert dialog = new Alert(Alert.AlertType.CONFIRMATION, "Quit SR2 Mod Manager?");
		ObservableList<ButtonType> dlgButtons = dialog.getDialogPane().getButtonTypes();
		dlgButtons.clear();
		dlgButtons.add(ButtonType.YES);
		dlgButtons.add(ButtonType.NO);
		dialog.setHeaderText("Quit SR2MM");

		dialog.showAndWait()
				.filter(response -> response == ButtonType.YES)
				.ifPresent(response -> System.exit(0));
	}

	@FXML private void deleteRepository(ActionEvent actionEvent) {
		Alert dialog = new Alert(Alert.AlertType.CONFIRMATION, "Are you sure you want to delete this repository from your computer? You will not be able to access it again until you redownload it.");
		ObservableList<ButtonType> dlgButtons = dialog.getDialogPane().getButtonTypes();
		dlgButtons.clear();
		dlgButtons.add(ButtonType.YES);
		dlgButtons.add(ButtonType.NO);
		dialog.setHeaderText("Delete Repository?");

		dialog.showAndWait()
				.filter(response -> response == ButtonType.YES)
				.ifPresent(response -> {
						File root = repo.getRepository().getWorkTree();
						repo.close();
						Utils.deleteFolder(root);

						if(root.exists()) {
							Alert msg = new Alert(Alert.AlertType.ERROR, "Could not delete repository!\n\nFor some reason, the repository was not deleted. You may have to delete it yourself.");
							msg.show();
						}
						else {
							branches.clear();
							branchList.getItems().clear();
							modInfo.setText("");
							branchInfo.setText("");
						}
				});
	}

	@FXML private void about(ActionEvent actionEvent) {
		Alert msg = new Alert(Alert.AlertType.INFORMATION,
				"SR2 Mod Manager by Dalo Lorn\n" +
						"\n" +
						"Version: 1.0.0 Beta 1\n\n" +
						"GitHub: https://github.com/DaloLorn/SR2ModManager\n" +
						"Discord: https://discord.gg/sUJKJDc\n" +
						"Patreon: https://patreon.com/rising_stars_sr2");
		msg.setHeaderText("About SR2 Mod Manager");
		msg.show();
	}
}

File: src/main/java/io/github/openstarruler/launchpad/view/MainController.kt
package io.github.openstarruler.launchpad.view

import io.github.openstarruler.launchpad.adapter.ModInstaller
import io.github.openstarruler.launchpad.adapter.Settings
import io.github.openstarruler.launchpad.adapter.TextHandler
import io.github.openstarruler.launchpad.adapter.Utils
import javafx.application.Platform
import javafx.concurrent.Task
import javafx.fxml.FXML
import javafx.geometry.Insets
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.control.Alert.AlertType
import javafx.scene.layout.VBox
import javafx.stage.DirectoryChooser
import javafx.stage.Modality
import javafx.stage.Stage
import javafx.stage.StageStyle
import java.io.File
import kotlin.system.exitProcess

class MainController {
    // Shared UI components.
    @FXML private lateinit var playButton: Button

    @FXML private lateinit var tabs: TabPane
    @FXML private lateinit var modsTab: Tab
    @FXML private lateinit var installerTab: Tab
    @FXML private lateinit var osrTab: Tab

    @FXML private lateinit var openRepositoryItem: MenuItem
    @FXML private lateinit var deleteRepositoryItem: MenuItem
    @FXML private lateinit var uninstallModItem: MenuItem

    @FXML private lateinit var modsPane: ModManagerPane
    @FXML private lateinit var installerPane: ModInstallerPane
    @FXML private lateinit var osrPane: OpenSRManagerPane

    fun initialize() {
        tabs.selectionModel.selectedItemProperty().addListener { _, _, newTab ->
            deleteRepositoryItem.isDisable = newTab !== installerTab || !ModInstaller.hasRepo()
            uninstallModItem.isDisable = newTab !== installerTab
        }
        osrPane.playButton = playButton
        installerPane.deleteRepositoryItem = deleteRepositoryItem

        if (
            Settings.instance.isFirstRun
        ) {
            val msg = ResizableAlert(
                AlertType.INFORMATION,
                "Welcome to the OpenSR Launchpad!\n\nAs this is your first time running the Launchpad, we probably don't know where you want to install Star Ruler 2 (or where it is already installed).\n\nIf you want to use an existing SR2 installation: Please navigate to the root folder of your SR2 installation, containing the 'data', 'locales', and 'scripts' folders.\n\nIf you want to install a fresh copy of OpenSR instead, please navigate to whatever folder you want to install the game into."
            )
            msg.showAndWait()
            Platform.runLater { setSR2Path() }
        }
        else if (
            !File(Settings.instance.gamePath, "Star Ruler 2.exe").exists()
            && !File(Settings.instance.gamePath, "StarRuler2.sh").exists()
        ) {
            val msg = ResizableAlert(
                AlertType.WARNING,
                "Could not detect Star Ruler 2 launchers! Please navigate to the root folder of your SR2 installation, containing the 'data', 'locales', and 'scripts' folders!"
            )
            msg.showAndWait()
            Platform.runLater { setSR2Path() }
        }
    }

    @FXML
    private fun setSR2Path() {
        osrPane.setSR2Path()
    }

    @FXML
    private fun openRepository() {
        // Get repository location.
        val chooser = DirectoryChooser()
        chooser.initialDirectory = File("repositories/")
        chooser.initialDirectory.mkdirs()
        chooser.title = "Choose a Repository"
        val dir = chooser.showDialog(playButton.scene.window)
            ?: return // Anything on the window would suffice, playButton was just arbitrarily selected.

        deleteRepositoryItem.isDisable = true
        executeTask("Loading repository...") { taskUpdateStage: Stage ->
            object : Task<Unit>() {
                override fun call() {
                    var repoURL: String
                    val errorHandler = TextHandler { text ->
                        Platform.runLater {
                            val msg: Alert = ResizableAlert(AlertType.ERROR, text)
                            msg.show()
                        }
                    }
                    if (ModInstaller.openRepository(
                            dir,
                            errorHandler
                        ).also { repoURL = it!! } == null
                    ) return
                    updateMessage("Fetching branch list...")
                    if (!installerPane.getBranches(errorHandler)) return
                    updateMessage("Getting mod description...")
                    if (!installerPane.getDescription(errorHandler)) return
                    Platform.runLater { installerPane.setUrlText("Connected to $repoURL") }
                    return
                }

                override fun succeeded() {
                    deleteRepositoryItem.isDisable = false
                    taskUpdateStage.close()
                }
            }
        }
    }

    @FXML
    private fun close() {
        val dialog = ResizableAlert(AlertType.CONFIRMATION, "Quit OpenSR Launchpad?")
        val dlgButtons = dialog.dialogPane.buttonTypes
        dlgButtons.clear()
        dlgButtons.add(ButtonType.YES)
        dlgButtons.add(ButtonType.NO)
        dialog.headerText = "Quit OpenSR Launchpad"
        dialog.showAndWait()
            .filter { response: ButtonType -> response == ButtonType.YES }
            .ifPresent { exitProcess(0) }
    }

    @FXML
    private fun deleteRepository() {
        if (installerTab.isSelected) {
            installerPane.deleteRepository()
        }
    }

    @FXML
    private fun about() {
        val msg: Alert = ResizableAlert(
            AlertType.INFORMATION,
            """
                  OpenSR Launchpad
                  
                  Version: 2.0.0
                  
                  GitHub: https://github.com/OpenSRProject/OpenStarRuler-Launchpad
                  Discord: https://discord.gg/sUJKJDc
                  """.trimIndent()
        )
        msg.headerText = "About OpenSR Launchpad"
        msg.show()
    }

    @FXML
    private fun startGame() {
        val launcher =
            File(Settings.instance.gamePath, if (Utils.IS_WINDOWS) "Star Ruler 2.exe" else "StarRuler2.sh")
        if (!launcher.exists()) ResizableAlert(
            AlertType.ERROR,
            "Cannot start Star Ruler 2: Couldn't find platform-appropriate loader!"
        )
            .showAndWait()
        if (!launcher.canExecute()) ResizableAlert(
            AlertType.ERROR,
            "Cannot start Star Ruler 2: Couldn't run game loader!"
        )
            .showAndWait()
        try {
            ProcessBuilder(launcher.absolutePath)
                .directory(launcher.parentFile)
                .start()
        } catch (e: Exception) {
            val msg = ResizableAlert(AlertType.ERROR, String.format("Cannot start Star Ruler 2: %s", e))
            e.printStackTrace()
            msg.showAndWait()
        }
    }

    @FXML
    fun uninstallMod() {
        if (installerTab.isSelected) {
            installerPane.uninstallMod()
        }
    }

    companion object {
        fun executeTask(
            initialText: String,
            taskBuilder: (Stage) -> Task<Unit>
        ) {
            // Start preparing the task.
            val wndwWidth = 800.0
            val taskLabel = Label(initialText)
            taskLabel.padding = Insets(0.0, 20.0, 0.0, 0.0)
            taskLabel.maxWidth = Double.MAX_VALUE
            taskLabel.isWrapText = true
            taskLabel.textOverrun = OverrunStyle.CENTER_ELLIPSIS

            val progress = ProgressBar()
            progress.maxWidth = Double.MAX_VALUE

            val taskPane = VBox()
            taskPane.padding = Insets(10.0, 10.0, 30.0, 10.0)
            taskPane.spacing = 5.0
            taskPane.children.addAll(taskLabel, progress)

            val taskStage = Stage(StageStyle.UTILITY)
            taskStage.initModality(Modality.APPLICATION_MODAL)
            taskStage.isResizable = true
            taskStage.scene = Scene(taskPane)
            taskStage.width = wndwWidth
            taskStage.show()

            val task = taskBuilder(taskStage)
            progress.progressProperty().bind(task.progressProperty())
            taskLabel.textProperty().bind(task.messageProperty())
            taskStage.show()
            Thread(task).start()
        }
    }
}
